fqcnt/fqcnt_rs1_rustbio.rs

```diff
@@ -0,0 +1,67 @@
+use bio::io::fastq;
+use flate2::bufread::MultiGzDecoder;
+use std::env;
+use std::fs::File;
+use std::io::BufReader;
+use std::path::{Path, PathBuf};
+
+/// A an "extension" trait to allow for extending the [`std::path::Path`](https://doc.rust-lang.org/nightly/std/path/struct.Path.html) struct.
+trait PathExt {
+    fn is_compressed(&self) -> bool;
+}
+
+impl PathExt for Path {
+    /// Determine if a `Path` is for a compressed file. This is based on whether the path ends with
+    /// the extension `.gz`.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// let path = std::path::Path::new("output.fq.gz");
+    ///
+    /// assert!(path.is_compressed())
+    /// ```
+    fn is_compressed(&self) -> bool {
+        match self.extension() {
+            Some(p) => p == "gz",
+            _ => false,
+        }
+    }
+}
+
+fn main() {
+    let path: PathBuf = match env::args().nth(1) {
+        Some(p) => PathBuf::from(p),
+        _ => {
+            eprintln!("Usage: {} <in.fq>", file!());
+            std::process::exit(1)
+        }
+    };
+
+    let file = match File::open(&path) {
+        Ok(fh) => fh,
+        Err(err) => {
+            eprintln!("Failed to open file: {}", err);
+            std::process::exit(1)
+        }
+    };
+
+    let file_handle = BufReader::new(file);
+
+    let stream: Box<dyn std::io::Read> = if path.is_compressed() {
+        Box::new(MultiGzDecoder::new(file_handle))
+    } else {
+        Box::new(file_handle)
+    };
+
+    let records = fastq::Reader::new(stream).records().map(|r| r.unwrap());
+    let mut n: u32 = 0;
+    let mut slen: u64 = 0;
+    let mut qlen: u64 = 0;
+    for record in records {
+        n += 1;
+        slen += record.seq().len() as u64;
+        qlen += record.qual().len() as u64;
+    }
```

Comment:
@mbhall88 Maybe using a pre-allocated mutable record might further improve the timing a bit. Would you mind trying that out with your setup?

For an example of this setup, checkout the FastqRead Trait docs: https://docs.rs/bio/0.30.1/bio/io/fastq/trait.FastqRead.html

Reply:
Not sure I follow sorry? Maybe you could illustrate with a PR?

Reply:
Yeah, sorry, that was jotted down a bit shoddily. I meant something like this:
```
    let mut reader = fastq::Reader::new(stream);
    let mut record = fastq::Record::new();

    let mut n: u32 = 0;
    let mut slen: u64 = 0;
    let mut qlen: u64 = 0;

    reader.read(&mut record)?;
    while !record.is_empty() {
        n += 1;
        slen += record.seq().len() as u64;
        qlen += record.qual().len() as u64;
        reader.read(&mut record)?
    }
```
A colleague of mine (@tedil) tested this locally and cut the time further. According to @johanneskoester it avoids re-allocating a new record on every iteration.

Reply:
Oh, cool! DIdn't realise that would have such a dramatic effect. For some reason, I assumed that Rust's Iterators would have just implicitly done the same thing when looping over them.

Reply:
https://github.com/lh3/biofast/pull/12